### PR TITLE
refactor: Introduce Flow Stages in order to deal with tests with unconventional booking flows

### PR DESCRIPTION
**Pull request by lukehesluke**
Purpose:

- Make it easy to write tests that have flows outside of the standard "fetch -> C1 -> C2 -> B" flow

With FlowStages, it should be possible to set up a Flow to use any stages in any order, including repeating some stages (e.g. "... -> p -> order feed update -> cancel -> order feed update").

----

This is currently unfinished. Remaining tasks:

1. Adapt it to a few more "awkward flow" tests to make sure that it can flexibly work with all these kinds of tests without issue
2. Add some helper functions to allow for being able to write tests in a much less verbose way. You can see that there's a bit of boilerplate now (like supplying `logger` to most FlowStages, specifying each pre-requisite even though there is a linear sequence, separate "initiator" and "collector" stages for order/opportunity feed updates)

Therefore, I ask that you review this as an underlying engine. However, I welcome any comments about developer experience, and I'll try to make the top layer with those in mind.

Though it is unfinished, I'd like to merge it into the code base to reduce the long-standing nature of this branch

---

I recommend starting with packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js, which has been adapted to use FlowStages

### CODE REVIEW

**Review by nickevansuk — COMMENTED** [review]
Just some ideas

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 1]
```diff
@@ -31,101 +30,68 @@ FeatureHelper.describeFeature(module, {
   // even if some OrderItems don't require approval, the whole Order should
   controlOpportunityCriteria: 'TestOpportunityBookable',
 },
-(configuration, orderItemCriteria, featureIsImplemented, logger, state, flow) => {
-  beforeAll(async () => {
-    await state.fetchOpportunities(orderItemCriteria);
+(configuration, orderItemCriteria, featureIsImplemented, logger) => {
+  // ## Initiate Flow Stages
+  const fetchOpportunities = FetchOpportunitiesFlowStage.create({ orderItemCriteria, logger });
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:35]: I did play about with this @nickevansuk. Great minds! It requires a little bit of figuring out for this test as one of the stage descriptions use facts of previous stages i.e. the action uses a thunk which uses `p.getResponse()`.

But it would be easy to make a function which does that stitching together for simpler tests which do not have requirements like the above.

And it would also be possible to make a more flexible one which supports the above case. I'll try to include something at a later stage of this PR

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 2]
```diff
@@ -31,101 +30,68 @@ FeatureHelper.describeFeature(module, {
...
-      .successChecks()
-      .validationTests();
-
-    itShouldReturnOrderRequiresApprovalTrue(() => state.c2Response);
+  const orderFeedUpdate = OrderFeedUpdateFlowStage.create({
+    testName: 'Order Feed (after Simulate Seller Approval)',
+    // orderFeedUpdate must be initiated by simulateSellerApproval because the
+    // order may appear on the feed as soon as simulateSellerApproval is
+    // completed.
+    initiatedBy: simulateSellerApproval,
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:54]: Great minds again! I totally agree. This is mostly a placeholder for now. I don't like that this approach means that some secret background mutation needs to occur, which might be surprising, but I do like that it means that the stages are written in order. The solution to [your other comment about using an array](https://github.com/openactive/openactive-test-suite/pull/215#discussion_r502764202) might have a way to have it both ways.

But I agree that in the less simple case, it should be the least magic version

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 3]
```diff
@@ -31,101 +30,68 @@ FeatureHelper.describeFeature(module, {
   // even if some OrderItems don't require approval, the whole Order should
   controlOpportunityCriteria: 'TestOpportunityBookable',
 },
-(configuration, orderItemCriteria, featureIsImplemented, logger, state, flow) => {
-  beforeAll(async () => {
-    await state.fetchOpportunities(orderItemCriteria);
+(configuration, orderItemCriteria, featureIsImplemented, logger) => {
+  // ## Initiate Flow Stages
+  const fetchOpportunities = FetchOpportunitiesFlowStage.create({ orderItemCriteria, logger });
+  const c1 = C1FlowStage.create({ preRequisite: fetchOpportunities, logger });
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:36]: I didn't realise that was one word. Okay!

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nickevansuk) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:35]: Another related thought, what if we used the chaining pattern here to set `prerequisite` and `initiatedBy` automatically? e.g.

```javascript
const flowSet = FlowFactory.createFlow(logger)
  .addStage('FetchOpportunities', FetchOpportunitiesFlowStage, { orderItemCriteria })
  .addStage('C1', C1FlowStage)
  .addStage('C2', C2FlowStage)
  .addStage('TestInterface`, TestInterfaceActionFlowStage, {
     testName: 'Simulate Seller Approval (Test Interface Action)',
     createActionFn: () => ({
       type: 'test:SellerAcceptOrderProposalSimulateAction',
       objectType: 'OrderProposal',
       objectId: p.getResult().body['@id'],
     })
  .addStage('OrderFeedUpdate', OrderFeedUpdateFlowStage, C2FlowStage)
  .addStage('B', BFlowStage)
```

That would make it slightly magic, however would mean we only need to reference the logger once, and different stages could then be accessed via the key in the resulting `Set`. We could even default the Key, so `  .addStage('C1', C1FlowStage)` could become `.addStage(C1FlowStage)`?

Might be more concise given it'll be repeated so many times? Might also provide more flexibility than the array pattern as params can be added to the stages.

File: packages/openactive-integration-tests/test/helpers/flow-stages/opportunity-feed-update.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  const relevantOffers = getRelevantOffers(opportunityCriteria, opportunity, { harvestStartTime: HARVEST_START_TIME });
+  if (relevantOffers.length === 0) { return null; }
+
+  return relevantOffers[Math.floor(Math.random() * relevantOffers.length)];
+}
+
+/**
+ * Wait for an opportunity to appear in the Booking System's Opportunity Feed.
+ */
+const OpportunityFeedUpdateFlowStage = {
```
Comment (lukehesluke) [inline packages/openactive-integration-tests/test/helpers/flow-stages/opportunity-feed-update.js:46]: I haven't set-up a `create(..)` function for this flow stage yet as this stage is not needed for the accept-proposal-book test.

However, the `run()` function has been coded as it is used by `fetch-opportunities`

File: packages/openactive-integration-tests/test/helpers/flow-stages/opportunity-feed-update.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+   * @param {object} args
+   * @param {ChakramResponse[]} args.testInterfaceOpportunities
+   * @param {OpportunityCriteria[]} args.orderItemCriteriaList
+   * @param {RequestHelperType} args.requestHelper
+   * @returns {Promise<import('./flow-stage').FlowStageOutput<{
+   *   opportunityFeedExtractResponses: ChakramResponse[],
+   *   orderItems: OrderItem[]
+   * }>>}
+   */
+  async run({ testInterfaceOpportunities, orderItemCriteriaList, requestHelper }) {
```
Comment (lukehesluke) [inline packages/openactive-integration-tests/test/helpers/flow-stages/opportunity-feed-update.js:57]: The body of this function has mostly been copied from RequestState.getMatch

File: packages/openactive-integration-tests/test/helpers/flow-stages/fetch-opportunities.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+ *
+ * The responses only contain minimal info like ID, so `runOpportunityFeedUpdate()`
+ * needs to be called afterwards to actually get the full data of each opportunity.
+ *
+ * @param {object} args
+ * @param {OpportunityCriteria[]} args.orderItemCriteriaList
+ * @param {RequestHelperType} args.requestHelper
+ * @returns {Promise<ChakramResponse[]>}
+ */
+async function getOrCreateTestInterfaceOpportunities({ orderItemCriteriaList, requestHelper }) {
```
Comment (lukehesluke) [inline packages/openactive-integration-tests/test/helpers/flow-stages/fetch-opportunities.js:38]: The code for this function has mostly been copied from RequestState.fetchOpportunities

File: packages/openactive-integration-tests/test/helpers/flow-stages/fetch-opportunities.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+          getterFn: () => flowStage.getResponse(),
+        });
+      },
+      itValidationTestsFn(flowStage) {
+        OpportunityFeedUpdateFlowStage.itValidationTests({
+          logger,
+          orderItemCriteriaList,
+          getterFn: () => flowStage.getResponse(),
+        });
+      },
```
Comment (lukehesluke) [inline packages/openactive-integration-tests/test/helpers/flow-stages/fetch-opportunities.js:119]: Note that this FlowStage doesn't use the same FlowStageUtils helper functions to generate success checks and validation tests.

That's because this FlowStage is a little more complicated - it doesn't just produce a single OpenActive item, but it produces a lot of data. This is why it has bespoke test functions

File: packages/openactive-integration-tests/test/helpers/request-helper.js [hunk 8]
```diff
@@ -266,9 +269,10 @@ class RequestHelper {
   /**
    * @param {string} uuid
    * @param {import('../templates/c1-req').C1ReqTemplateData} params
-   * @param {import('../templates/c1-req').C1ReqTemplateRef} c1ReqTemplateRef
+   * @param {import('../templates/c1-req').C1ReqTemplateRef | null} [maybeC1ReqTemplateRef]
    */
-  async putOrderQuoteTemplate(uuid, params, c1ReqTemplateRef = 'standard') {
+  async putOrderQuoteTemplate(uuid, params, maybeC1ReqTemplateRef) {
+    const c1ReqTemplateRef = maybeC1ReqTemplateRef || 'standard';
```
Comment (lukehesluke) [inline packages/openactive-integration-tests/test/helpers/request-helper.js:275]: This change means that a `null` is also converted to the default value

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js]: @nickevansuk I'm going to resolve this for now as I'm going to do the top-level magic function in a separate PR

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js]: @nickevansuk in the end I split it into 2 FlowStages, which is very non-magic (quite boilerplate even!) - an "initiator" stage and a "collector" stage. This means that the core FlowStage logic doesn't need special affordances for this kind of use case (initiating feed updates before the prerequisite stage).

However, my plan is to have it so that the magic function on top will abstract away these details so that you don't have to think about creating 2 FlowStages.

Therefore, it's rather boilerplate-y now, but that's just the "low-level" code, which will generally be abstracted away. Lmkwyt

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 9]
```diff
@@ -31,101 +38,90 @@ FeatureHelper.describeFeature(module, {
...
+      objectType: 'OrderProposal',
+      objectId: p.getResponse().body['@id'],
+    }),
+    requestHelper,
+  });
+  const collectOrderFeedUpdate = OrderFeedUpdateFlowStage.createCollector({
+    testName: 'Order Feed Update (after Simulate Seller Approval)',
+    prerequisite: simulateSellerApproval,
+    logger,
+  });
```
Comment (lukehesluke) [inline packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:83]: See the documentation for OrderFeedUpdateFlowStage to understand what these 2 "initiate" and "collect" stages are all about.

**Review by nickevansuk — COMMENTED** [review]
So a few specific comments on the example test, and also more general points below:

It took me a while to figure out how state was being passed around, so it's not super intuitive at the moment... however it is powerful when understood, so perhaps that's a good tradeoff and just needs some good docs.

There are quite a few benefits to the prereq approach to state passing vs. having explicit input and output from each stage. Though I was reflecting on whether just having e.g. getC1Output being passed into C2 would make it clearer what's going on, it would also make all the tests more verbose for the sake of a minor edge case...

There's a dependency tree that's implicitly formed, so we probably want to make clear that the prereq that's specified is valid (e.g. of the correct type?). Could we use TS to do this somehow? e.g. for "B" and "P" prereq must be C2, for "B" prereq must be either "C2" or "collectOrderFeedUpdate"? This also ensures that if someone is testing C1 -> C1 -> C2 -> C2 -> B -> B (e.g. for idempotency) they are using proper prereq (and not e.g. using B as a prereq for B)

There's also a question about how to make the tests easy to reason about when each stage transforms state from the previous stage... which could lead to some interesting bugs. 

For example, tests such as C1 -> C2 -> P -> feedCheck -> Simulate -> feedCheck -> Simulate -> feedCheck -> B -> feedCheck (for testing approval states) could be difficult to reason about, especially if there's an error somewhere in there.

Some (incomplete) thoughts on this:
- Perhaps each stage should only add state. Hence could we key the state by stage, so that it’s clear what’s getting added where, and what each stage is depending on? E.g. "c2.totalPaymentDue" could be added to the state by C2, and read by B or P? If we enforce that a stage cannot prereq itself (as per point above) this creates an invariant where the state for each stage is immutable outside of the stage that created it.
- This actually also has the added benefit of making the test flow independent of the data returned - e.g for C1, C2(i), C2(ii), B this would ensure that B would need to either use the results of C2(i) OR C2(ii), it could not be set up so that the results of either are used.
- Additionally there are also some tests where we want to test that both C2 and B return errors (e.g. due to no spare capacity), so need to ensure that B will still run even if there's an error returned by C2 (and hence the state of e.g. totalPaymentDue cannot be set).
- For situations like the above, and others, do we also need a way of transforming/overriding state at a particular stage. For example we may want to test an incorrect "totalPaymentDue" by having B use "totalPaymentDue from C2 + 1"? To test invalid orderProposalVersion might also want to just use a dummy orderProposalVersion for B?
- State transforms probably then solve most usecases where we need to amend something prereq data... do we also want to allow a run of e.g. B with completely fabricated C1 and C2 (they're not even run). I guess a pre-run state transform at B could allow the test to just insert values into the state that would otherwise have been expected from C1 and C2? This could be used to test, e.g. UnknownOrderError at Order Deletion endpoint (where only the Order Deletion endpoint is used, but the UUID needs to be fabricated)?
- Also the above approach works for repeating calls to endpoints (which shouldn't appear twice in a dependency chain), and it works for repeated calls for feed updates (the only place an Orders feed check is required as a prereq is orderProposalVersion for B). E.g. the test "Run C1, C2 and B, then run U (prereq of B), then check Orders feed (prereq of B), then run Order Deletion (prereq of B), and check Orders feed again (prereq of B) that Order has been deleted".  So perhaps Orders feed only being a prereq for B means we we don't need to solve the problem of how do have the Orders feed collection run a second time without mutating the state from the first Orders feed collection?

Having written all the above, I've also realised that the prereq is also used to define the order in which calls are made, and we want to ensure that e.g. an Orders Feed collection completes before the next call to B is made. So perhaps immutable state per-stage is not a simple solution after all...

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 10]
```diff
@@ -31,101 +38,90 @@ FeatureHelper.describeFeature(module, {
...
+  const b = BFlowStage.create({
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:84]: > If implicit, should this be more explicit?

I agree, however I think it's outside of the scope of this PR as that's how the logic was before also (not sure it's worth an issue but I'll keep a note)

**Comment by lukehesluke** [comment]
@nickevansuk I am very interested in your 1st suggestion to use something like getC1Output (`() => ChakramResponse`).

I would propose slightly more generic names like those in the FlowStageState type e.g. `getBookingSystemOrder` to facilitate the wide range of flows (e.g. B's param could be getC2Output() or getPOutput() or getOrderFeedUpdateOutput(). Or all 3?).

The current approach has an assumption which is limiting to some tests - it gets the result from the last stage that produced a booking system order. What if we want to write a test that does "C1 -> C2 -> P -> B (using the data from C1 rather than from C2 or P)". Like you say, this assumption is some behind-the-scenes magic.

Most of the other questions you've raised demonstrate that this suggestion would simplify all those use cases.

I think this framework should be very flexible so that all possible use cases are supported (because it's so simple). Then, we can have "magic" functions that reduce some of the flexibility for the price of reduced boilerplate. This way, all flows are supported and only as verbose as they need to be.

**Comment by lukehesluke** [comment]
Here are some more specific responses. I stopped half-way through when it became apparent that the 1st suggestion (explicit getter function passed in to each FlowStage) answers all of them

- > There's a dependency tree that's implicitly formed, so we probably want to make clear that the prereq that's specified is valid (e.g. of the correct type?). Could we use TS to do this somehow? e.g. for "B" and "P" prereq must be C2, for "B" prereq must be either "C2" or "collectOrderFeedUpdate"? 
    - This should be possible with TS yes. I didn't want to add too many limitations to be honest, because it might not support every test (e.g. where we intentionally do stages in the wrong place or repeatedly).
- > This also ensures that if someone is testing C1 -> C1 -> C2 -> C2 -> B -> B (e.g. for idempotency) they are using proper prereq (and not e.g. using B as a prereq for B)
    - I'm not sure I understand what you mean. By my understanding, your earlier suggestion that a prereq must be of a set range of types would prohibit the ability to do a test like C1 -> C1 -> C2 -> C2 -> B -> B (e.g. you'd get a TS error "C1 cannot have C1 as a prerequisite")
- > Perhaps each stage should only add state. Hence could we key the state by stage, so that it’s clear what’s getting added where, and what each stage is depending on? E.g. "c2.totalPaymentDue" could be added to the state by C2
    - I initially decided against this as it seems that it limits the possible flows. e.g. what about "C1 -> B". Even though this should fail at the Booking System, it will fail in the code because there's no state in `c2`.  Additionally, it raises questions like "should B use the state from C2 or P? Should it be P and then C2 if there's nothing at P because that's the usual order? What if a test has subverted that order?".
- > e.g for C1, C2(i), C2(ii), B this would ensure that B would need to either use the results of C2(i) OR C2(ii), it could not be set up so that the results of either are used.
    - I don't think I understand. If state was keyed such that C2 always updates `c2.`, wouldn't that cause a potential issue in the above example where the state would only be taken from C2 (ii), as it is run last (of the C2s).

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 11]
```diff
@@ -31,101 +38,90 @@ FeatureHelper.describeFeature(module, {
...
-      createActionFn: () => ({
-        type: 'test:SellerAcceptOrderProposalSimulateAction',
-        objectType: 'OrderProposal',
-        objectId: state.pResponse.body['@id'],
-      }),
-      completedFlowStage: 'P',
-    }))
-      .beforeSetup()
-      .successChecks();
+  const initiateOrderFeedUpdate = OrderFeedUpdateFlowStage.createInitiator({
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js]: (this is done)

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 12]
```diff
@@ -31,101 +38,90 @@ FeatureHelper.describeFeature(module, {
...
+  FlowStageUtils.describeRunAndCheckIsSuccessfulAndValid(fetchOpportunities);
 
-    it('should have orderProposalStatus: SellerAccepted', () => {
-      expect(orderFeedUpdate.getStateResponse().body).to.have.nested.property('data.orderProposalStatus', 'https://openactive.io/SellerAccepted');
-    });
-    it('should have orderProposalVersion same as that returned by P (i.e. an amendment hasn\'t occurred)', () => {
-      expect(orderFeedUpdate.getStateResponse().body).to.have.nested.property('data.orderProposalVersion', state.pResponse.body.orderProposalVersion);
-    });
+  FlowStageUtils.describeRunAndCheckIsSuccessfulAndValid(c1, {
+    itExtraTests() {
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js]: (this is done)

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 13]
```diff
@@ -1,7 +1,14 @@
 const { expect } = require('chai');
 const { FeatureHelper } = require('../../../../helpers/feature-helper');
-const { GetMatch, C1, C2, P, OrderFeedUpdate, TestInterfaceAction, B } = require('../../../../shared-behaviours');
-// const { GetMatch, C1, C2, P } = require('../../../../shared-behaviours');
+const { FetchOpportunitiesFlowStage } = require('../../../../helpers/flow-stages/fetch-opportunities');
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js]: (this is done)

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js [hunk 14]
```diff
@@ -31,101 +40,101 @@ FeatureHelper.describeFeature(module, {
...
+  FlowStageUtils.describeRunAndCheckIsSuccessfulAndValid(fetchOpportunities);
 
-    it('should have orderProposalStatus: SellerAccepted', () => {
-      expect(orderFeedUpdate.getStateResponse().body).to.have.nested.property('data.orderProposalStatus', 'https://openactive.io/SellerAccepted');
-    });
-    it('should have orderProposalVersion same as that returned by P (i.e. an amendment hasn\'t occurred)', () => {
-      expect(orderFeedUpdate.getStateResponse().body).to.have.nested.property('data.orderProposalVersion', state.pResponse.body.orderProposalVersion);
-    });
+  FlowStageUtils.describeRunAndCheckIsSuccessfulAndValid(c1, {
+    itAdditionalTests() {
```
Comment (nickevansuk) [inline packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:107]: Just realised this should be called `additionalTests` not `itAdditionalTests`, as `itAdditionalTests` implies it has a built-in `it()` (as is the case in other places), where in fact the `it` must be included within the definied `additionalTests` function body.

**Review by nickevansuk — CHANGES_REQUESTED** [review]
One minor naming point so far...

File: packages/openactive-integration-tests/test/helpers/flow-stages/flow-stage.js [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+ * @template {FlowStageOutput} TInput
+ * @template {FlowStageOutput} TOutput
+ */
+class FlowStage {
+  /**
+   * @param {object} args
+   * @param {FlowStage<unknown, unknown>} [args.prerequisite] Stage that must be completed before
+   *   this stage is run. e.g. a C2 stage might have a C1 stage as its
+   *   pre-requisite.
+   * @param {() => TInput} args.getInput TODO TODO document
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/helpers/flow-stages/flow-stage.js]: fixed

File: packages/openactive-integration-tests/test/helpers/flow-stages/flow-stage.js [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+ * @property {ChakramResponse[]} [testInterfaceOpportunities] Opportunity
+ *   references for opportunities used by this flow. These match the response
+ *   schema for the POST /test-interface/datasets/:testDatasetIdentifier/opportunities
+ *   (https://openactive.io/test-interface/#post-test-interfacedatasetstestdatasetidentifieropportunities)
+ *   endpoint.
+ * @property {ChakramResponse[]} [opportunityFeedExtractResponses]
+ * @property {OrderItem[]} [orderItems]
+ * @property {unknown} [bookingSystemOrder] Order as found in the Booking
+ *   System. Could contain the response from C1, C2, B, etc.
+ *   TODO TODO TODO is this field actually being used anywhere?
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/helpers/flow-stages/flow-stage.js]: fixed

File: packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js (hunk 14 above)
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:107]: @nickevansuk my understanding of the convention is:

Any function that begins with `it` spawns `it()` blocks. Therefore, this usage is correct, as `itAdditionalTests` is a function which is expected to spawn `it()` blocks

Comment (nickevansuk) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:107]: Ah that's not the case at least in these examples...

https://github.com/openactive/openactive-test-suite/blob/95efec494df99ab237efe0aaa2166f4ff055a70b/packages/openactive-integration-tests/test/shared-behaviours/errors.js#L14-L21

https://github.com/openactive/openactive-test-suite/blob/348adcb3d55db809276cac6885b7ae07625aab19/packages/openactive-integration-tests/test/shared-behaviours/common.js#L40-L58

https://github.com/openactive/openactive-test-suite/blob/348adcb3d55db809276cac6885b7ae07625aab19/packages/openactive-integration-tests/test/shared-behaviours/common.js#L24-L27

Comment (lukehesluke) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:107]: Those functions' names begin with `it` and they spawn `it()` blocks - like the `itAdditionalTests` functions.

I don't think I understand

Comment (nickevansuk) [reply packages/openactive-integration-tests/test/features/approval/minimal-proposal/implemented/accept-proposal-book-test.js:107]: Or perhaps more specifically, the `it*` function is not expected to contain a parameter that requires further `it` blocks... it is itself an `it` block, as the `describe*` functions are describe blocks...

**Review by nickevansuk — APPROVED** [review]
